common/models/templates/src/tokenization.ts

```diff
@@ -45,75 +96,103 @@ export function tokenize(wordBreaker: WordBreakingFunction, context?: Partial<Co
+
+  // Note:  is MOSTLY "WET" with the left-span loop, though the
+  // `caretSplitsToken` check is additional.
+  while(rightSpans.length > 0) {
+    const nextSpan = rightSpans[0];
+    if(nextSpan.start != currentIndex) {
+      // Implicit whitespace span!
+      tokenization.right.push({
+        text: context.right!.substring(currentIndex, nextSpan.start),
+        isWhitespace: true
+      });
+      currentIndex = nextSpan.start;
+    } else {
+      // If the first non-whitespace token to the right is non-whitespace,
+      // and the last token to the left is non-whitespace, the caret may
+      // be splitting a token.
+      const leftTail = tokenization.left[leftTokenCount-1];
+      if(firstRightToken && !leftTail.isWhitespace) {
+        if(wordBreaker(leftTail!.text + nextSpan.text).length == 1) {
+          tokenization.caretSplitsToken = true;
+        }
+      }
 
-    if(leftSuffixWordbreak || rightPrefixWordbreak) {
-      // Bypass the final test, as we already know the caret didn't split a token.
-      // (The tokenization process already removed characters between the two.)
-      return tokenization;
+      // Explicit non-whitespace span.
+      rightSpans.shift();
+      tokenization.right.push({
+        text: nextSpan.text
+      });
+      currentIndex = nextSpan.end;
     }
 
-    // Worth note - some languages don't use wordbreaking characters.  So, a final check:
-    //
-    // Does the wordbreaker split a merge of the 'two center' tokens?
-    // If not, then the caret is responsible for the split.
-    if(wordBreaker(leftTail!.text + rightHead.text).length == 1) {
-      tokenization.caretSplitsToken = true;
-    }
+    // We've always processed the "first right token" after the first iteration.
+    // Do not run the caret-split check on any future iterations.
+    firstRightToken = false;
+  }
+
+  // Detect any pre-caret whitespace after the final pre-caret non-whitespace
+  // token
+  //
+  // Note:  the default wordbreaker won't need this code, as it emits a `''`
+  // after final whitespace.
+  //
+  // Also note:  is pretty much WET with the similar check after the
+  // leftSpan loop.
+  if(currentIndex != (context.right?.length ?? 0)) {
+    tokenization.right.push({
+      text: context.right!.substring(currentIndex, context.right!.length),
+      isWhitespace: true
+    });
+    currentIndex = context.right!.length;
   }
 
   return tokenization;
 }
 
 /**
  * Get the last word of the phrase before the caret or nothing.
+ * If the last 'token' before the caret is whitespace, returns `''`.
  * @param fullLeftContext the entire left context of the string.
  */
 export function getLastPreCaretToken(wordBreaker: WordBreakingFunction, context: Context): string {
   let tokenization = tokenize(wordBreaker, context);
   if (tokenization.left.length > 0) {
-    return tokenization.left.pop() as string;
+    const lastToken = tokenization.left.pop();
+    if(lastToken!.isWhitespace) {
```

Comment:
Again, do we need this assertion? I'd prefer to see code where we either test for nullishness or we can prove it to be non-nullish.